cdap-ui/app/cdap/components/LogViewer/index.tsx

```diff
@@ -0,0 +1,387 @@
+    const currentScrollHeight = logsContainer.scrollHeight;
+
+    this.props.dataFetcher.getPrev().subscribe(
+      (res) => {
+        if (res.length > 0) {
+          const newLogs = res.concat(this.state.logs);
+          this.setState({ logs: newLogs });
+        }
+
+        this.setState({ isFetching: false }, () => {
+          // maintaining scroll position
+          const updatedScrollHeight = logsContainer.scrollHeight;
+          logsContainer.scrollTop = updatedScrollHeight - currentScrollHeight;
+
+          const trimmedLogs = this.state.logs.slice(0, MAX_LOG_ROWS);
+
+          // trim excess logs
+          this.setState({
+            logs: trimmedLogs,
+          });
+          this.props.dataFetcher.onLogsTrim(
+            trimmedLogs[0],
+            trimmedLogs[this.state.logs.length - 1]
+          );
+        });
+      },
+      (err) => {
+        // tslint:disable-next-line: no-console
+        console.log('err', err);
+      }
+    );
+  }, 100);
+
+  private fetchNext = debounce(() => {
+    if (this.state.isFetching || this.state.isPolling) {
+      return;
+    }
+    this.setState({ isFetching: true });
+
+    this.props.dataFetcher.getNext().subscribe(
+      (res) => {
+        if (res.length > 0) {
+          const newLogs = this.state.logs.concat(res);
+          this.setState({ logs: newLogs });
+        }
+
+        this.setState({ isFetching: false }, this.trimTopExcessLogs);
+      },
+      (err) => {
+        // tslint:disable-next-line: no-console
+        console.log('err', err);
+      }
+    );
+  }, 100);
+
+  // Trimming the top entries of logs will change the scroll height. This function will calculate the
+  // delta of the scroll height to maintain the view.
+  private trimTopExcessLogs() {
+    const logsContainer = this.logsContainer.current;
+    const offset = logsContainer.scrollHeight - logsContainer.scrollTop;
+
+    if (this.state.logs.length > MAX_LOG_ROWS) {
+      this.setState(
+        {
+          logs: this.state.logs.slice(this.state.logs.length - MAX_LOG_ROWS),
+        },
+        () => {
+          logsContainer.scrollTop = logsContainer.scrollHeight - offset;
+          this.scrollPosition = logsContainer.scrollTop;
+
+          this.props.dataFetcher.onLogsTrim(
+            this.state.logs[0],
+            this.state.logs[this.state.logs.length - 1]
+          );
+        }
+      );
+    }
+  }
+
+  private cleanUpWatchers = () => {
+    this.io.disconnect();
+    this.stopScrollWatch();
+    this.stopPoll();
+  };
+
+  private getLatestLogs = () => {
+    this.cleanUpWatchers();
+
+    this.setState(
+      () => {
+        return {
+          isFetching: true,
+          isPolling: true,
+        };
+      },
+      () => {
+        this.props.dataFetcher.getLast().subscribe(this.processFirstResponse);
+      }
+    );
+  };
+
+  private changeLogLevel = (level: LogLevelEnum) => {
+    this.cleanUpWatchers();
+
+    this.setState(
+      () => {
+        return {
```

Comment:
nit: Any specific reason why the setState is using a callback function instead of directly setting the state?

Reply:
i was trying to fix a problem, and I thought it was because of React micro batching setState, but it was not. So this can be reverted